### PR TITLE
New event endpoint

### PR DESCRIPTION
**Pull request by AbdurRafay**
* Payload change

* Event Builder Test Cases modified and added for new payload

* Test cases modified for uuid

* Added assertion failure messages

### CODE REVIEW

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m [hunk 1]
```diff
@@ -21,16 +21,24 @@
 NSString * const OPTLYEventParameterKeysRevision                        = @"revision";
 NSString * const OPTLYEventParameterKeysActivationId                    = @"activationId";
 NSString * const OPTLYEventParameterKeysSessionId                       = @"sessionId";
```
Comment (kellyroach-optimizely) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m]: Optimizely python-sdk EventParams doesn't mention "session_id" and I do not
see OPTLYEventParameterKeysSessionId acutally being used anywhere in the
objective-c-sdk .  Therefore, I think you can delete this line.

**Review by kellyroach-optimizely — CHANGES_REQUESTED** [review]

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m [hunk 2]
```diff
@@ -21,16 +21,24 @@
...
-NSString * const OPTLYEventParameterKeysVisitorUUID                     = @"visitorUUID";
-NSString * const OPTLYEventParameterKeysProjectId                       = @"projectId";         // nonnull
-NSString * const OPTLYEventParameterKeysAccountId                       = @"accountId";         // nonnull
-NSString * const OPTLYEventParameterKeysClientEngine                    = @"clientEngine";
-NSString * const OPTLYEventParameterKeysClientVersion                   = @"clientVersion";
+NSString * const OPTLYEventParameterKeysVisitorId                       = @"visitor_id";        // nonnull
+NSString * const OPTLYEventParameterKeysUUID                            = @"uuid";
+NSString * const OPTLYEventParameterKeysProjectId                       = @"project_id";        // nonnull
+NSString * const OPTLYEventParameterKeysAccountId                       = @"account_id";        // nonnull
+NSString * const OPTLYEventParameterKeysClientEngine                    = @"client_name";
```
[the remark this answers is not in the document]
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m]: Yes I agreed. but with reference to other SDKs like (c sharp, php, ruby) they are still using ClientEngine as the parameter key name, so for consistency I also left it as it is.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m (hunk 1 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m]: OPTLYEventParameterKeysSessionId removed.

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYEventBuilderTest.m [hunk 3]
```diff
@@ -563,67 +607,67 @@ - (void)testRevenueMetricAndValueMetric
...
+//                  eventTags:@{OPTLYEventMetricNameRevenue:@(kEventRevenue)}];
+//    NSArray *experimentIds = @[@"6364835526", @"6450630664", @"6367863211", @"6376870125", @"6383811281", @"6358043286", @"6370392407", @"6367444440", @"6370821515", @"6447021179"];
+//    NSArray *layerStates = params[OPTLYEventParameterKeysLayerStates];
+//    NSUInteger numberOfLayers = [layerStates count];
+//    NSUInteger numberOfExperiments = [experimentIds count];
+//    // 6383811281 (testExperimentWithFirefoxAudience) is excluded because the attributes do not match
+//    // 6367444440 (testExperimentNotRunning) is excluded because the experiment is not running
+//    // 6450630664 should be exlucded becuase it is mutually excluded.
+//    XCTAssert(numberOfLayers == (numberOfExperiments - 3), @"Incorrect number of layers.");
+//}
```
[the remark this answers is not in the document]
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYEventBuilderTest.m]: Sorry I forgot to remove it as there are no event metrics and layer states in the payload to test it. 
removing now..

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYEventBuilderTest.m [hunk 4]
```diff
@@ -563,67 +607,67 @@ - (void)testRevenueMetricAndValueMetric
...
+//    [self commonBuildConversionTicketTest:@{kAttributeKeyBrowserType:kAttributeValueChrome,
+//                                       @"IntegerTag":@15,
+//                                       @"BooleanTag":@YES,
+//                                       @"FloatTag":@1.23,
+//                                       @"InvalidArrayTag":[NSArray new]}
+//                       sentEventTags:@{kAttributeKeyBrowserType:kAttributeValueChrome,
+//                                       @"IntegerTag":@15,
+//                                       @"FloatTag":@1.23,
+//                                       @"BooleanTag":@YES}];
+//}
```
[the remark this answers is not in the document]
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYEventBuilderTest.m]: Uncommenting it to test the restrictions of event tags value data types.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (kellyroach-optimizely) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m]: OK, for this P.R., we can let this slip.  Later, we should fix all the SDKs.

**Comment by kellyroach-optimizely** [comment]
I would like this pull request to also delete these OPTLYEventParameterKeys.m+.h symbols:
    * "OPTLYEventParameterKeysActivationId"
    * "OPTLYEventParameterKeysUserFeatures"
    * "OPTLYEventParameterKeysIsGlobalHoldback"
    * "OPTLYEventParameterKeysMetricName"
According to the Avro *.avsc specification I've seen, these aren't used.
According to the Objective-C code I've seen, these aren't used in any significant way
in code.

**Comment by kellyroach-optimizely** [comment]
ADDITIONAL ITEMS IN OPTLYEventParameterKeys.m+.h THAT SHOULD BE REMOVED

Purpose of the P.R. is reorg of the JSON event information sent to new V3 endpoint including event batching, drift towards snake_case keys away from camelCase keys, and some moves described in JIRA OASIS-1867 .  Granting this, there should be no vestiges of "eventEntityId", "eventName", nor "eventFeatures" left around in the objective-c-sdk code, but there are still such vestiges.

We believe these items in OPTLYEventParameterKeys.m+.h should be removed:

```
NSString * const OPTLYEventParameterKeysEventEntityId                   = @"eventEntityId";
NSString * const OPTLYEventParameterKeysEventType                       = @"eventType";
NSString * const OPTLYEventParameterKeysEventName                       = @"eventName";
NSString * const OPTLYEventParameterKeysEventFeatures                   = @"eventFeatures";     // nonnull
NSString * const OPTLYEventParameterKeysEventMetrics                    = @"eventMetrics";      // nonnull
NSString * const OPTLYEventParameterKeysRelatedEvents                   = @"relatedEvents";
NSString * const OPTLYEventParameterKeysLayerStates                     = @"layerStates";       // nonnull
NSString * const OPTLYEventParameterKeysActiveViews                     = @"activeViews";
NSString * const OPTLYEventParameterKeysRelatedEventsEventId            = @"eventId";
NSString * const OPTLYEventParameterKeysRelatedEventsRelationship       = @"relationship";
```

**Comment by kellyroach-optimizely** [comment]
REGARDING @"eventId" CLEANUP

The fate of @"eventId" is a bit more complicated.  However, if we
search the "reference" python-sdk for either of "eventId" or "event_id", we don't find these.
Instead, we find "entity_id" in python-sdk's optimizely/event_builder.py
```
   EVENT_ID = 'entity_id'
```

**Comment by AbdurRafay** [comment]
All unreferenced items of OPTLYEventParameterKeys.m+.h has beed removed and reordered them to their respective object.

**Comment by kellyroach-optimizely** [comment]
REMAINING OCCURRENCES OF LITERAL STRINGS @"eventEntityId" ETC. SHOULD BE GONE TOO

Most of the literal string offenders are in OptimizelySDKShared/OptimizelySDKSharedTests/OPTLYDataStoreTest.m .
These grep commands in the source code directory should be producing empty output:
```
grep -r \"eventEntityId\" .
grep -r \"eventName\" .
grep -r \"eventFeatures\" .
grep -r \"eventMetrics\" .
grep -r \"layerStates\" .
```

Footnote to the above: Mixpanel code is third party, is fine, and don't change it
when addressing @"eventName" .  At first glance, the keyMapper code in
OptimizelySDKCore/OptimizelySDKCore/OPTLYEvent.m might appear suspect,
but actually it is fine, and we don't want to change this code either.

**Comment by kellyroach-optimizely** [comment]
TEST:
* sh ./Scripts/build_all.sh PASSED
* sh ./Scripts/local_travis.sh PASSED
* 2 iOS & 2 tvOS Clean / Build / Analyze PASSED
* 8 iOS test suites PASSED
* OptimizelySDKiOSUniversalTests PASSED
* 6 tvOS test suites PASSED but 1 tvOS test suite FAILED
- (void)testTVOSSDKInitializedWithOverrides
* OptimizelySDKTVOSUniversalTests PASSED
* iOS & tvOS DemoApp's PASSED

The one test failure in OptimizelySDKTVOSTests "- (void)testTVOSSDKInitializedWithOverrides"
looks rather minor, is not due to P.R. #202, and I or someone else will see to getting that
test to pass in some future P.R.  There is a similar at risk test in OptimizelySDKiOSTests.m which is just getting lucky up to now.

CONCLUDE: We're going to APPROVE the P.R. #202 .

**Review by kellyroach-optimizely — APPROVED** [review]

**Comment by kellyroach-optimizely** [comment]
This was OASIS-1867 .

**Pull request merged**